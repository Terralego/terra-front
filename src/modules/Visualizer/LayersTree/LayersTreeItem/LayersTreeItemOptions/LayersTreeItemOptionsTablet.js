import React from 'react';
import classnames from 'classnames';
import { Button } from '@blueprintjs/core';

import translateMock from '../../../../../utils/translate';

import LayerFetchValues from '../LayerFetchValues';
import FiltersPanel from '../FiltersPanel';

const LayersTreeItemOptionsTablet = ({
  hasSomeOptionActive,
  isOptionsOpen,
  handleOptionPanel,
  layer,
  toggleFilters,
  isFilterVisible,
  getFilterPanelRef,
  form,
  toggleTable,
  isTableActive,
  displayTableButton,
  toggleWidgets,
  widgets,
  isWidgetActive,
  translate = translateMock({
    'terralego.visualizer.layerstree.itemOptions.widget.title': 'widget',
    'terralego.visualizer.layerstree.itemOptions.widget.action-open': 'open widget',
    'terralego.visualizer.layerstree.itemOptions.widget.action-close': 'close widget',
    'terralego.visualizer.layerstree.itemOptions.widget.title_synthesis': 'widget synthesis',
    'terralego.visualizer.layerstree.itemOptions.widget.action-open_synthesis': 'open synthesis',
    'terralego.visualizer.layerstree.itemOptions.widget.action-close_synthesis': 'close synthesis',
    'terralego.visualizer.layerstree.itemOptions.table.label': 'table',
    'terralego.visualizer.layerstree.itemOptions.table.alt': 'open table',
    'terralego.visualizer.layerstree.itemOptions.table.alt_close': 'close table',
    'terralego.visualizer.layerstree.itemOptions.filter.label': 'filters',
    'terralego.visualizer.layerstree.itemOptions.filter.alt': 'open filters',
    'terralego.visualizer.layerstree.itemOptions.filter.alt_close': 'close filters',
    'terralego.visualizer.layerstree.itemOptions.options.label': 'options',
    'terralego.visualizer.layerstree.itemOptions.options.alt': 'open options',
    'terralego.visualizer.layerstree.itemOptions.options.alt_close': 'close options',
  }),
}) => (
  <div
    className={classnames(
      'layerstree-node-content__options',
      'layerstree-node-content__options--mobile',
      { 'layerstree-node-content__options--active': hasSomeOptionActive },
    )}
  >
    {(widgets && !!widgets.length) && (
      // i18next-extract-mark-context-start ["", "synthesis"]
      widgets.map(widget => {
        const { component: context } = widget;
        const isActive = isWidgetActive(widget);

        const actionText = isActive
          ? translate('terralego.visualizer.layerstree.itemOptions.widget.action-close', { context })
          : translate('terralego.visualizer.layerstree.itemOptions.widget.action-open', { context });
        return (
          <Button
            key={context}
            className={classnames({
              'btn-widget': true,
              'layerstree-node-content__options__button': true,
              'layerstree-node-content__options__button--active': isWidgetActive(widget),
            })}
            onClick={toggleWidgets(widget)}
            minimal
            icon="selection"
            title={translate('terralego.visualizer.layerstree.itemOptions.widget.title', {
              context,
            })}
            alt={actionText}
          >
            {context}
          </Button>
        );
      })
      // i18next-extract-mark-context-stop
    )
}
    {displayTableButton && (
    <Button
      className={classnames(
        'btn-table',
        'layerstree-node-content__options__button',
        { 'layerstree-node-content__options__button--active': isTableActive },
      )}
      onClick={toggleTable}
      minimal
      icon="th"
      title={translate('terralego.visualizer.layerstree.itemOptions.table.title')}
      alt={translate('terralego.visualizer.layerstree.itemOptions.table.alt', {
        context: isTableActive ? 'close' : 'open',
      })}
    >
      {translate('terralego.visualizer.layerstree.itemOptions.table.label')}
    </Button>
    )}
    {form && (
      <FiltersPanel
        visible={isFilterVisible}
        onMount={getFilterPanelRef}
        layer={layer}
      >
        {isFilterVisible && (
          <LayerFetchValues layer={layer} isFilterVisible={isFilterVisible} />
        )}
        <Button
          className={
          classnames(
            'btn-form',
            'layerstree-node-content__options__button',
            { 'layerstree-node-content__options__button--active': isFilterVisible },
          )}
          onClick={toggleFilters}
          minimal
          icon="filter"
          title="filter"
        >
        filtres
        </Button>
      </FiltersPanel>
    )}
    <Button
      className={classnames(
        'btn-option',
        'layerstree-node-content__options__button',
        'layerstree-node-content__options__button--more',
        { 'layerstree-node-content__options__button--active': isOptionsOpen },
      )}
      icon="more"
      minimal
      onClick={handleOptionPanel}
      title={translate('terralego.visualizer.layerstree.itemOptions.options.label')}
      alt={translate('terralego.visualizer.layerstree.itemOptions.options.alt', {
        context: isOptionsOpen ? 'close' : 'open',
      })}
    >
      {translate('terralego.visualizer.layerstree.itemOptions.options.label')}
    </Button>
  </div>
);
export default LayersTreeItemOptionsTablet;
