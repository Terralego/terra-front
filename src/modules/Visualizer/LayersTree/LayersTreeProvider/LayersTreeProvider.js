import React from 'react';
import PropTypes from 'prop-types';
import { connectState } from '../../../State/context';
import { TYPE_RANGE } from '../../../Forms/Filters';
import { serialize } from '../../../../utils/serialize';

import context from './context';
import {
  initLayersStateAction,
  setLayerStateAction,
} from '../../services/layersTreeUtils';

const { Provider } = context;

export class LayersTreeProvider extends React.Component {
  static propTypes = {
    /** Callback executed everytime layersTreeState change. Takes layersTreeState as parameter */
    onChange: PropTypes.func,
    /** Initial state */
    initialState: PropTypes.shape({
      layers: PropTypes.oneOfType([PropTypes.string, PropTypes.array]), // Active layer id(s)
      table: PropTypes.string, // Layer if table is active
    }),
    /** Initial layer tree state */
    initialLayersTreeState: PropTypes.instanceOf(Map),
    /**
     * Function called when a filter property of single or many type need to fetch values
     * Takes `layer` and `property` as parameters
     * @return String[]
     * */
    fetchPropertyValues: PropTypes.func,
    /**
     * Function called when a filter property of range type need to fetch min and max
     * Takes `layer` and `property` as parameters
     * @return {{}} Object of form {min: Number, max: Number}
     * */
    fetchPropertyRange: PropTypes.func,
    translate: PropTypes.func,
    setCurrentState: PropTypes.func,
  };

  static defaultProps = {
    onChange () {},
    initialState: {},
    initialLayersTreeState: new Map(),
    fetchPropertyValues () {},
    fetchPropertyRange () {},
    translate: undefined,
    setCurrentState () {},
  };

  constructor (props) {
    super(props);
    const { initialLayersTreeState: layersTreeState } = this.props;
    this.state = { layersTreeState };
  }

  componentDidMount () {
    this.initLayersState();
  }

  componentDidUpdate ({ initialLayersTreeState: prevLayersTreeState, layersTree: prevLayersTree }) {
    const { initialLayersTreeState, layersTree } = this.props;

    if (initialLayersTreeState !== prevLayersTreeState) {
      // Deep comparison to avoid unusefull and buggy rerender.
      // This is a hack waiting better refactoring.
      if (serialize(initialLayersTreeState) !== serialize(prevLayersTreeState)) {
        this.initLayersState(initialLayersTreeState);
      }
    }

    if (layersTree !== prevLayersTree) {
      this.initLayersState();
    }
  }

  componentWillUnmount () {
    this.isUnmount = true;
  }

  setLayerState = ({ layer, state: newState, reset }) => {
    this.resetState(({ layersTreeState }) => ({
      layersTreeState: setLayerStateAction(layer, newState, layersTreeState, reset),
    }));
  };

  getLayerState = ({ layer }) => {
    const { layersTreeState } = this.state;
    return layersTreeState.get(layer) || {};
  };

  fetchPropertiesValues = async (layer, properties) => {
    const { fetchPropertyValues, fetchPropertyRange } = this.props;
    const { layersTreeState } = this.state;
    properties.forEach(property => {
      // We need to keep a static reference to this object because it serve as
      // key in layersTreeState
      // eslint-disable-next-line no-param-reassign
      property.loading = true;
      if (property.type) {
        // eslint-disable-next-line no-param-reassign
        property.values = [];
      }
    });
    this.resetState(new Map(layersTreeState));

    const responses = await Promise.all(properties.map(property =>
      (property.type === TYPE_RANGE
        ? fetchPropertyRange(layer, property)
        : fetchPropertyValues(layer, property))));

    properties.forEach((property, index) => {
      const { type } = property;
      /* eslint-disable no-param-reassign */
      delete property.loading;
      if (type === TYPE_RANGE) {
        const { min = 0, max = 100 } = responses[index] || {};
        property.min = min;
        property.max = max;
      } else {
        property.values = [...(responses[index] || [])];
      }
      /* eslint-enable no-param-reassign */
    });

    const { layersTreeState: newLayersTreeState } = this.state;
    this.resetState(new Map(newLayersTreeState));
  };

  initLayersState = initialLayersTreeState => {
    this.resetState(({ layersTreeState }) => {
      const { layersTree, initialState } = this.props;
      const state = initialLayersTreeState || layersTreeState;

      if (!layersTree) return {};

      return {
        layersTreeState: state.size
          ? state
          : initLayersStateAction(layersTree, initialState),
      };
    });
  };

  /**
   * Reset the tree state.
   *
   * @param {Object|Function} state
   * @param {Function} callback
   */
  resetState (state, callback = () => { }) {
    const { setCurrentState, onChange } = this.props;

    const setStateCallback = () => {
      callback();
      const { layersTreeState = new Map() } = this.state;

      // Simplify the state from the map
      const activeLayers = [];
      let table;

      const populateActiveLayersAndTable = (layerState, { layers: [layerId] = [] }) => {
        if (layerState.active) {
          activeLayers.push(layerId);
        }
        if (layerState.table) {
          table = layerId;
        }
      };

      layersTreeState.forEach(populateActiveLayersAndTable);
      setCurrentState({ layers: activeLayers, table });
      onChange(layersTreeState);
    };

    this.setState(state, setStateCallback);
  }

  render () {
    const {
      children,
      layersTree,
      map,
      translate,
    } = this.props;
    const { layersTreeState } = this.state;
    const {
      initLayersState, setLayerState, getLayerState,
      fetchPropertiesValues,
    } = this;
    const value = {
      map,
      layersTree,
      layersTreeState,
      initLayersState,
      setLayerState,
      getLayerState,
      fetchPropertiesValues,
      translate,
    };
    return (
      <Provider value={value}>
        {children}
      </Provider>
    );
  }
}

export default connectState('initialState', 'setCurrentState')(LayersTreeProvider);
