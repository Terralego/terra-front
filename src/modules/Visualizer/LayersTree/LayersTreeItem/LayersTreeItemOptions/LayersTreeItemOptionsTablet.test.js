import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import LayersTreeProvider from '../../LayersTreeProvider';
import LayersTreeItemOptionsTablet from './LayersTreeItemOptionsTablet';

jest.mock('../LayerFetchValues', () => () => <p>LayerFetchValues</p>);
jest.mock('../FiltersPanel', () => (({ children }) => children));

jest.mock('../FiltersPanel', () => function FiltersPanel () {
  return <p>FiltersPanel</p>;
});

it('should render correctly with widget', () => {
  const tree = renderer.create((
    <LayersTreeProvider>
      <LayersTreeItemOptionsTablet
        isFilterVisible
        form={[]}
      />
      <LayersTreeItemOptionsTablet
        isFilterVisible
        form={[]}
        isWidgetActive={() => true}
        isOptionsOpen={() => true}
        widgets={[{ component: 'foo' }]}
        toggleWidgets={() => null}
      />
    </LayersTreeProvider>
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should open widget', () => {
  const wrapper = shallow((
    <LayersTreeItemOptionsTablet
      displayTableButton
      form={[]}
      widgets={[{ component: 'foo' }]}
      isWidgetActive={() => 'close'}
      toggleWidgets={() => true}
    />
  ));
  expect(wrapper.find('.btn-widget').props().alt).toBe('close widget');
});

it('should close widget', () => {
  const wrapper = shallow((
    <LayersTreeItemOptionsTablet
      displayTableButton
      form={[]}
      widgets={[{ component: 'foo' }]}
      isWidgetActive={() => 'open'}
      toggleWidgets={() => true}
    />
  ));
  expect(wrapper.find('.btn-widget').props().alt).toBe('open widget');
});

it('should close table', () => {
  const wrapper = shallow((
    <LayersTreeItemOptionsTablet
      displayTableButton
      isTableActive
    />
  ));
  expect(wrapper.find('.btn-table').props().alt).toBe('close table');
});
