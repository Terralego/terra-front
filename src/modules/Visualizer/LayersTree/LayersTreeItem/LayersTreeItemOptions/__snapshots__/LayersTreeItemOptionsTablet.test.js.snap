// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with widget 1`] = `
Array [
  <div
    className="layerstree-node-content__options layerstree-node-content__options--mobile"
  >
    <p>
      FiltersPanel
    </p>
    <button
      alt="open options"
      className="bp3-button bp3-minimal btn-option layerstree-node-content__options__button layerstree-node-content__options__button--more"
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      title="options"
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-more"
        icon="more"
      >
        <svg
          data-icon="more"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            more
          </desc>
          <path
            d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        className="bp3-button-text"
      >
        options
      </span>
    </button>
  </div>,
  <div
    className="layerstree-node-content__options layerstree-node-content__options--mobile"
  >
    <button
      alt="close widget"
      className="bp3-button bp3-minimal btn-widget layerstree-node-content__options__button layerstree-node-content__options__button--active"
      onClick={null}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      title="widget"
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-selection"
        icon="selection"
      >
        <svg
          data-icon="selection"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            selection
          </desc>
          <path
            d="M8 0C3.58 0 0 3.58 0 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-9C6.34 5 5 6.34 5 8s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        className="bp3-button-text"
      >
        foo
      </span>
    </button>
    <p>
      FiltersPanel
    </p>
    <button
      alt="close options"
      className="bp3-button bp3-minimal btn-option layerstree-node-content__options__button layerstree-node-content__options__button--more layerstree-node-content__options__button--active"
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      title="options"
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-more"
        icon="more"
      >
        <svg
          data-icon="more"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            more
          </desc>
          <path
            d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <span
        className="bp3-button-text"
      >
        options
      </span>
    </button>
  </div>,
]
`;
